### PR TITLE
fix(federation): fixed unnecessary sibling typename aliasing

### PR DESCRIPTION
**Pull request by duckki**
In JS version, Field's `siblingTypename` field can be either undefined, empty string or an alias name.
In Rust, the `Name` type can't hold empty string and the empty string case wasn't implemented before. This PR fixed that by adding a new enum type to distinguish the unaliased/aliased cases.

One previously failing snapshot test is now fixed.
Many mismatches from corpus data are fixed.

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [x] Integration Tests
    - [ ] Manual Tests

### CODE REVIEW

File: apollo-federation/src/query_plan/fetch_dependency_graph.rs [hunk 1]
```diff
@@ -3325,18 +3325,13 @@ fn compute_nodes_for_op_path_element<'a>(
...
-            None
-        } else {
-            Some(name.clone())
-        };
+        let alias = sibling_typename.alias();
         let typename_field = Arc::new(OpPathElement::Field(Field::new_introspection_typename(
             &dependency_graph.supergraph_schema,
             &operation.parent_type_position(),
-            alias,
+            alias.cloned(),
```
Comment (goto-bus-stop) [inline apollo-federation/src/query_plan/fetch_dependency_graph.rs:3334]:
```suggestion
            sibling_typename.alias().cloned(),
```
We don't really need the `alias` binding here anymore

File: apollo-federation/src/operation/mod.rs [hunk 2]
```diff
@@ -2683,16 +2707,11 @@ impl SelectionSet {
...
-                None
-            } else {
-                Some(sibling_typename.clone())
-            };
+            let alias = sibling_typename.alias();
             let field_element = Field::new_introspection_typename(
                 &self.schema,
                 &selection.element()?.parent_type_position(),
-                alias,
+                alias.cloned(),
```
Comment (goto-bus-stop) [inline apollo-federation/src/operation/mod.rs:2714]:
```suggestion
            sibling_typename.alias().cloned(),
```
We don't really need the `alias` binding here anymore

**Review by goto-bus-stop — APPROVED** [review]
In theory we could've used `name == name!(__typename)` conditions instead of `name.is_empty()`, but I like that the enum forces you to deal with it!

**Comment by duckki** [comment]
> In theory we could've used `name == name!(__typename)` conditions instead of `name.is_empty()`, but I like that the enum forces you to deal with it!

I thought about something like that. But, enum would be clearer and avoid future bugs like this.

**Comment by dariuszkuc** [comment]
> In Rust, the Name type can't hold empty string and the empty string case wasn't implemented before. This PR fixed that by adding a new enum type to distinguish the unaliased/aliased cases.

Whats the use case for empty string? I was under impression that we either have an alias (which has to be a valid GraphQL name) or we don't (so its None).

**Comment by duckki** [comment]
> > In Rust, the Name type can't hold empty string and the empty string case wasn't implemented before. This PR fixed that by adding a new enum type to distinguish the unaliased/aliased cases.
> 
> Whats the use case for empty string? I was under impression that we either have an alias (which has to be a valid GraphQL name) or we don't (so its None).

The `siblingTypename` field has three cases (in JS):

* `undefined` means the field is not tagged (so, no `__typename` sibling to be added).
* The empty string means the field should add a `__typename` sibling without an alias.
* A non-empty string means the field should add a `__typename` sibling with the string as its alias.

File: apollo-federation/src/operation/mod.rs [hunk 3]
```diff
@@ -1330,14 +1330,32 @@ mod field_selection {
         }
     }
 
+    // SiblingTypename indicates how the sibling __typename field should be restored.
+    // PORT_NOTE: The JS version used the empty string to indicate unaliased sibling typenames.
+    // Here we use an enum to make the distinction explicit.
+    #[derive(Debug, Clone)]
+    pub(crate) enum SiblingTypename {
```
Comment (dariuszkuc) [inline apollo-federation/src/operation/mod.rs:1337]: if we are using custom enum to represent the sibling typename -> maybe we should drop the `Option`? so we would have `None/Unaliased/Aliased`?

Comment (dariuszkuc) [reply apollo-federation/src/operation/mod.rs:1337]: Technically we could still use just `Option<Name>` without enum -> ie just initialize it with `__typename`

Comment (duckki) [reply apollo-federation/src/operation/mod.rs:1337]: Using a special name as non-alias would throw someone off at some point. So, I didn't want to do that.

I did consider `None/Unaliased/Aliased`, but I liked the added `Option` after all. First, the sibling typename handling logic for Unaliased/Aliased will take the same branch anyways to add `__typename`. The added `fn alias()` getter function, which returns `Option<Name>`, makes no further case-split necessary. After all, that's just a subtle difference. So, I took the path of least resistance, not changing all the existing `None` cases :)

**Pull request merged**